Add new blog post `Introducing the latest addition Action Validate` (#152)
conforma/conforma.github.io: website/content/posts/introducing-action-validate-image.md

---
title: "Introducing Action Validate for GitHub"
date: 2023-10-24T13:02:00-04:00
author: "Sean Conroy"
---
<!--more-->

Introducing Action Validate you may already be familiar with using the `EC-CLI Validate` command for local container image validation. Now, you can seamlessly integrate this functionality directly into your build processes or any other automated workflow in GitHub.

## What is EC Validate Action?

**EC Validate** is a GitHub Action that brings validation capabilities of the Enterprise Contract CLI directly into your GitHub Workflow. Designed to integrate seamlessly into your build pipeline, this action offers both compliance assessments and integrity checks for your container images. It ensures that your images meet both organizational and industry standards before deployment or environment promotion.

### Key Features

- **GitHub Native**: Being a GitHub Action, EC Validate seamlessly integrates into your existing GitHub workflows, while also providing GitHub summary output.
- **Policy Compliance**: Ability to tailor its validation checks based on custom or pre-defined policies.
- **Integrity Checks**: Verifies that the image hasn't been tampered with. 

> - **[Signature Support](https://enterprisecontract.dev/docs/ec-cli/main/signing.html)**: Offers support for both Long-Lived Public-Key Authentication and Keyless Authentication.

Reviewer: I don't think we need all the capitalization here.

Also some wording change suggestions.

```suggestion
- **[Signature Verification Support](https://enterprisecontract.dev/docs/ec-cli/main/signing.html)**: Offers support for verifying both long-lived public-key signed, and keyless signed container images.

- **`image`**: Specifies the container image to be validated. It uses the image and digest output from a previous build step in the workflow.

- **`identity`**: Defines the identity or entities that are expected to have produced the image signature and attestations. In this case, it's either from the SLSA GitHub Generator or the repository owner.

- **`issuer`**: Specifies the OIDC issuer of the token used for authentication. Here, it's set to GitHub's token actions issuer URL.

- **`policy`**: Points to the policy configuration to use for validation checks. The policy can be either [predefined](https://github.com/enterprise-contract/config) or a custom policy.

### Promote Image
Upon successful validation, this step promotes the image by pushing the latest validated tag. This ensures that only images that have passed checks are promoted. 
```yaml
- name: Push latest-validated image tag 
  run: |
      skopeo copy \
      --dest-creds=${{ github.actor }}:${{ github.token }} \       
      docker://${{ needs.build.outputs.image }}@${{ needs.build.outputs.digest } \     
      docker://${{ needs.build.outputs.image }}:latest
```

## Example of Long-Lived Public-Key Authentication

> Here is a version of the EC Action Validate that uses long-lived. This method uses a public key, stored in a secret variable, to verify the image signature, thereby ensuring its integrity through a three-stage validation process: Signature Verification, Attestation Verification, and Policy Compliance.

Reviewer: ```suggestion
Here is a version of the EC Action Validate that verifies artifacts signed by cosign with long-lived signing secrets. This method uses a public key, stored in a secret variable, to verify the image signature, thereby ensuring its integrity through a three-stage validation process: Signature Verification, Attestation Verification, and Policy Compliance.
```

```yaml
- name: Validate image (long-lived)
  uses: enterprise-contract/action-validate-image@v1.0.31 
  with:
    image: quay.io/redhat-appstudio/ec-golden-image:latest
    key: ${{ vars.PUBLIC_KEY }}
    policy: github.com/enterprise-contract/config//slsa3
    extra-params: --ignore-rekor
```

- **`image`**: Similar to keyless, specifies the container image to be validated.
- **`key`**: The public key used for long-lived authentication.
- **`policy`**: Policy configuration, which can be either [predefined](https://github.com/enterprise-contract/config) or custom.
- **`extra-params`**: Additional parameters for the action, such as ignoring Rekor for this image. More can be found [here](https://enterprisecontract.dev/docs/ec-cli/main/ec_validate_image.html#_options)

By using either keyless or long-lived authentication methods, you can tailor EC Action Validate to meet the specific security requirements of your project.

## Putting it all together

> EC Validate is a GitHub Action aimed at elevating the security and compliance of your container images right within your GitHub workflow. By offering a range of authentication methods and customizable or defined policies, this action ensures that only validated and compliant images make it to deployment. It's a solution for organizations looking to adhere to organizational and industry standards while automating their pipelines within GitHub.

Reviewer: Nice.

---

> title: "Introducing the latest addition Action Validate"

Reviewer: Needs a colon I think:

```suggestion
title: "Introducing the latest addition: Action Validate"
```

..but I'd probably go with something shorter that mentions GitHub

```suggestion
title: "Introducing Action Validate for GitHub"
```

> date: 2023-10-23T13:02:00-04:00

Reviewer 2: Let's add an `author` tag here :smiley_cat:

author: "Sean Conroy"
---
<!--more-->

> Introducing Action Validate

Reviewer: Not sure if this is formatted right. Is it supposed to be this maybe?
```suggestion
## Introducing Action Validate

```

Author: @simonbaird `date: 2023-04-25T12:56:35-04:00` was the problem updated it today without changing the time.

> Introducing [Action Validate](https://github.com/marketplace/actions/ec-validate).

Reviewer 2: This sentence is a bit odd. Maybe turn into something like: This blog introduces Action Validation.

But maybe just get rid of it altogether, and just rely on the link mentioned later on?

> and its `ec validate` command for local container image validation. Now, you can seamlessly integrate 

Reviewer 2: ... image validation, you can now seamlessly ...

I don't think the first sentence is a full sentence as it is. Alternatively, you can get rid of "while" and it should work as its own sentence.

> this functionality directly into your build processes or any other automated workflows in GitHub.

Reviewer 2: "any other automated workflow" (singular form)?

> <!--more-->

Reviewer 2: What's this for?

Author: I seen it in `introducing-the-enterprise-contract.md` guessing its not needed then.

Reviewer: This does have a purpose. The placement of `<!--more-->` marks the end of the excerpt which will show up on the home page. If you don't include it you'll get a default excerpt which might be too long or cut off mid sentence.

So maybe put it around line 10 or so.

> ## What is EC Action Validate?

Reviewer 2: "EC Action Validate" sounds a bit odd. I think if there's enough context, "EC Validate" is enough. Otherwise, the full name can be "EC Validate GitHub action", or just "EC Validate action (if there's enough context).

If changing here, let's change the other occurrences in the doc as well.

> **EC Action Validate** is a GitHub Action that brings validation capabilities of Enterprise Contract's EC CLI directly into your GitHub Workflow. Designed to integrate seamlessly into your build pipeline, this action offers both compliance assessments and intergrity checks for your container images. It ensures that your images meet both organizational and industry standards before deployment.

Reviewer 2: "of Enterprise Contract's EC CLI" -> "of the Enterprise Contract CLI"

> **EC Action Validate** is a GitHub Action that brings validation capabilities of Enterprise Contract's EC CLI directly into your GitHub Workflow. Designed to integrate seamlessly into your build pipeline, this action offers both compliance assessments and intergrity checks for your container images. It ensures that your images meet both organizational and industry standards before deployment.

Reviewer 2: Maybe change "before deployment" to "before deployment or environment promotion" ?

("before deployment" appears later on in the docs as well. If changing here, let's be consistent.)

### Key Features

> - **GitHub Native**: Being a GitHub Action, EC Action Validate seamlessly integrates into your existing GitHub workflows, while also provoiding github summary output.

Reviewer 2: Let's be consistent with how `github` is written. Stick to "GitHub" (avoid "github")?

- **Policy Compliance**: ability to tailor its validation checks based on custom or pre-defined policies.
- **Integrity Checks**: Verifies that the image hasn't been tampered with. 

> - **Authentication Support**: Offers support for both Long-Lived Public-Key Authentication and Keyless Authentication.

Reviewer 2: I don't know if I'd call this authentication. Maybe "Signature Support"?

Maybe linking here helps: https://enterprisecontract.dev/docs/ec-cli/main/signing.html ?

Author: Signature Support sounds a lot better 🥇

> Interested in learning more? Visit our [EC Action Market Place](https://github.com/marketplace/actions/ec-validate) for a user guide.

Reviewer 2: "Visit the EC Validate action in GitHub's Market Place" ?

## The `Golden-Container` Build Pipeline Example using EC Action Validate 
Imagine you're operating within a [build & release workflow](https://github.com/enterprise-contract/golden-container/blob/main/.github/workflows/release.yaml). You've successfully set up a pipeline that takes care of building, digitally signing, generating an SBOM, and adding provenance data. However, you're missing a crucial step validating the container image. This is where the `EC Action Validate` comes into play, ensuring your container images meet the required security and compliance standards before deployment.

> 

Reviewer 2: I think this needs a sentence that says you're going to walk-through the workflow before actually divining into it.

Author: ```suggestion 
Now, let’s go over the initial steps of this workflow before actually divining into it.
```

Reviewer: "diving" not "divining" unless I'm missing something.. :crystal_ball:

 ### Sign image with GitHub OIDC Token
In this step, the container image is signed using Cosign and a GitHub OIDC token. This adds an additional layer of security and trust to the image, making it easier to verify its integrity and origin.

```yaml
- name: Sign image with GitHub OIDC Token
  run: cosign sign --yes ${IMAGE_REGISTRY}/${IMAGE_REPO}@${DIGEST}
  env:
    DIGEST: ${{ steps.push-image.outputs.digest }}
```

### Generate and Store SBOM

> We use Syft to generate a SBOM. It's then attested using Cosign, which serves as both documentation and an extra security measure.

Reviewer 2: "documentation"?

    digest: ${{ needs.build.outputs.digest }}
    registry-username: ${{ github.actor }}
  secrets:
    registry-password: ${{ secrets.GITHUB_TOKEN }}
```

### EC Action Validate
EC Action Validate works by assessing your container images against a set of validation checks. You can customize these checks through a policy to align with the specific security and compliance guidelines of your organization or industry. Whether the image passes or fails, you'll receive a GitHub summary output, and additional logs will be available in YAML format within the action.
```yaml
- name: Validate image (keyless)
  uses: enterprise-contract/action-validate-image@v1.0.31 
  with:
    image: ${{ needs.build.outputs.image }}@${{ needs.build.outputs.digest }}
    identity: https:\/\/github\.com\/(slsa-framework\/slsa-github-generator|${{ github.repository_owner }}\/${{ github.event.repository.name }})\/
    issuer: https://token.actions.githubusercontent.com
    policy: github.com/enterprise-contract/config//github-default
```
- **`image`**: Specifies the container image to be validated. It uses the image and digest output from a previous build step in the workflow.

> - **`identity`**: Defines the identity or entities that are allowed to perform the validation. In this case, it's either from the SLSA GitHub Generator or the repository owner.

Reviewer 2: "... that are expected to have produced the image signature and attestations." ?

> - **`issuer`**: Specifies the issuer of the token used for authentication. Here, it's set to GitHub's token actions issuer URL.

Reviewer 2: "Specifies the OIDC issuer." ?

> - **`policy`**: Points to the policy configuration to use for validation checks. The policy can be either [predefined](https://github.com/enterprise-contract/config) or custom and is stored in a specific repository.

Reviewer 2: "and is stored in a specific repository" seems out of place here.

Author: Agreed, 'and is stored in a specific repository' might be confuse a reader. I'll simplify it to just 'or custom policy ' to keep it clear.

### Promote Image
Upon successful validation, this step promotes the image by pushing the latest validated tag. This ensures that only images that have passed checks are promoted. 
```yaml
- name: Push latest-validated image tag 

>   run: |

Reviewer 2: Either this needs to be `run: >` or you need `\` at then end of the two lines below.

      skopeo copy --dest-creds=${{ github.actor }}:${{ github.token }}       
      docker://${{ needs.build.outputs.image }}@${{ needs.build.outputs.digest }}      
      docker://${{ needs.build.outputs.image }}:latest
```

> To see the full workflow vist [here](https://github.com/enterprise-contract/golden-container/blob/main/.github/workflows/release.yaml)

Reviewer 2: It's already mentioned at the beginning of this section. Maybe not needed again?

## Example of Long-Lived Public-Key Authentication

> Here is a verison of the EC Action Validate that uses long-lived public-key authentication. This method uses a public key to authenticate the image, which is stored in a GitHub repository. This key is then used to verify the image's signature and ensure its integrity.

Reviewer 2: "authentication" terminology here is not right.

Author: ```suggestion
Here is a verison of the EC Action Validate that uses long-lived. This method uses a public key, stored in a secret variable, to verify the image signature, thereby ensuring its integrity through a three-stage validation process: Signature Verification, Attestation Verification, and Policy Conformance which we have usen `slsa3`.
```

---
title: "Introducing Action Validate for GitHub"
date: 2023-10-24T13:02:00-04:00
author: "Sean Conroy"
---

> Introducing Action Validate you may already be familiar 

Reviewer: This still seems like maybe it was supposed to be a heading..?

```suggestion
## Introducing Action Validate

You may already be familiar 
```

Or should that part be removed entirely?

```suggestion
You may already be familiar 
```

Reviewer: Apart from that, everything else looks lgtm. Nice work.

Reviewer: I'm thinking it's probably the second one, since there's already the main doc heading right above, so two similar headings might be overkill.
